pallets/capacity/src/lib.rs

```diff
@@ -430,26 +649,45 @@ impl<T: Config> Pallet<T> {
 		staker: &T::AccountId,
 		target: MessageSourceId,
 		amount: BalanceOf<T>,
+		staking_type: StakingType,
 	) -> Result<(StakingDetails<T>, BalanceOf<T>), DispatchError> {
 		ensure!(amount > Zero::zero(), Error::<T>::ZeroAmountNotAllowed);
 		ensure!(T::TargetValidator::validate(target), Error::<T>::InvalidTarget);
 
-		let staking_account = Self::get_staking_account_for(&staker).unwrap_or_default();
+		let staking_details = Self::get_staking_account_for(&staker).unwrap_or_default();
+		if !staking_details.active.is_zero() {
+			ensure!(
+				staking_details.staking_type.eq(&staking_type),
+				Error::<T>::CannotChangeStakingType
+			);
+		}
+
 		let stakable_amount = Self::get_stakable_amount_for(&staker, amount);
 
 		ensure!(stakable_amount > Zero::zero(), Error::<T>::BalanceTooLowtoStake);
 
-		let new_active_staking_amount = staking_account
+		let new_active_staking_amount = staking_details
 			.active
 			.checked_add(&stakable_amount)
 			.ok_or(ArithmeticError::Overflow)?;
 
 		ensure!(
 			new_active_staking_amount >= T::MinimumStakingAmount::get(),
-			Error::<T>::InsufficientStakingAmount
+			Error::<T>::StakingAmountBelowMinimum
 		);
 
-		Ok((staking_account, stakable_amount))
+		Ok((staking_details, stakable_amount))
+	}
+
+	fn ensure_can_boost(
+		staker: &T::AccountId,
+		target: &MessageSourceId,
+		amount: &BalanceOf<T>,
+	) -> Result<(StakingDetails<T>, BalanceOf<T>), DispatchError> {
+		let (mut staking_details, stakable_amount) =
+			Self::ensure_can_stake(staker, *target, *amount, StakingType::ProviderBoost)?;
+		staking_details.staking_type = StakingType::ProviderBoost;
```

Comment:
I’m not sure MaximumCapacity is the best default choice since it might not be the most common. If the default isn’t right for all situations, it could cause bugs or unexpected behavior, especially if developers assume it’s the right choice without realizing the potential issues.

https://github.com/frequency-chain/frequency/pull/1694/files#diff-4446b08e1f533bd6e73c4e3ee2ef514e8a7ab7dc63d770aaafcdc5a5d14bf596R74

Reply:
MaximumCapacity needs to be default for backward compatibility.

Reply:
This seems like an easy fix if we doing a migration. It appears that we can explicitly set it.

Reply:
On testnet at least a migration of that scale would need to be multiblock

Reply:
It sounds like you’ve already encountered an issue where, if a migration wasn’t completed in a previous PR, we end up with three states: Empty, ProviderBoost, and MaximumCapacity. This seems incorrect and could be impacting data integrity, potentially leading to unwanted results.